documentation/p1950.md

```diff
@@ -39,58 +32,42 @@ public:
-
-```
-// Header file
-class widget {
-public:
-    widget();
-    ~widget();
-
-    widget(const widget&);
-    widget& operator=(const widget&);
-private:
-    class impl;
-    std::experimental::propagate_const<std::unique_ptr<impl>> pimpl;
-};
- 
-// Implementation file
-class widget::impl {
-    // :::
-};
+### Issues with `unique_ptr<impl>` implementation of PImpl Idiom
+#### const-propagation
 
-widget::widget() : pimpl{ std::make_unique<impl>() } {}
-widget::~widget() = default;
+Using <code>std::unique_ptr</code> to store the implementation object introduces an issue: within a `const`-qualified member function, an instance of the visible class can mutate data inside the implementation object.  This is because <code>const</code> qualification applies only to the `unique_ptr` value, and not the pointed-to-object.  
 
-widget::widget(const widget& rhs)
-{ if (rhs.pimpl) { pimpl = std::make_unique<impl>(*pimpl); }   }
+The compiler is unable to give thread-compatibility guarantees for `const` objects when
+`const` is not propagated - `const` does not mean immutable in the face of pointer-like-member data.
 
-widget& widget::operator=(const widget& rhs)
-{
-    if (!rhs.pimpl ) { get_underlying(pimpl).reset(); }
-    else if (!pimpl ) { std::make_unique<impl>(*rhs.pimpl); }
-    else { *pimpl = *rhs.pimpl; }
+The desired semantics of a PImpl-owner are value-like, like those of `std::optional` which has appropriate `const`- and non-`const`-qualified overloads
+for `operator*` and `operator->`. 
 
-    return *this;
-}
+#### copies
+The copy-constructor and copy-assignment operator of `std::unique_ptr` are deleted. Users of a class with a `std::unique_ptr` member will be required to implement copy-construction and copy-assignment.  Implementation of these methods for a class with PImpl members is non-trivial [S. Meyers].
 
-```
-However, this is boilerplate code that must be repeated per Pimpl implementation.  Worse still we are not able to benefit from the rule-of-zero to have the compiler generate the methods for use.
+### Issues with `polymorphic_value<impl>` implementation of PImpl Idiom
+The class template `polymorphic_value` confers value-like semantics on a free-store allocated object. This satisfies our
+requirement for const-propagation and compiler-generated copy constructors but incurs an additional cost in supporting polymorphism.
```

Comment:
Prefer “a compiler generated copy constructor and copy assignment operator”

Reply:
Done